lib/Containers/Enumerate.h

```diff
@@ -0,0 +1,130 @@
+struct enumerate_wrapper;
+
+/*
+ * enumerate function similar to pythons enumerate.
+ * Usage:
+ *  for (auto [idx, e] : enumerate(v)) {
+ *    // idx is incremented for each element
+ *    // e is the reference type of the iterator of the container v.
+ *  }
+ *
+ * Usable for std::vector, std::list, std::map and std::unordered_map. _should_
+ * work for other containers that provide a LegacyIterator as well.
+ */
+template <typename T, typename C = unsigned int>
+enumerate_wrapper<T, C> enumerate(T&& v, C c = C());
+
+
+
+template <typename T, typename C = unsigned int>
+struct enumerate_iterator;
+
+template <typename T, typename C>
+void swap(enumerate_iterator<T, C>& a, enumerate_iterator<T, C>& b);
+
+template <typename T, typename C>
+struct enumerate_iterator {
+  enumerate_iterator() : _iter(), _c(0) {}
+
+  explicit enumerate_iterator(T iter, C c = C()) : _iter(iter), _c(c) {}
+  enumerate_iterator(enumerate_iterator const&) = default;
+  enumerate_iterator(enumerate_iterator&&) noexcept = default;
+
+  enumerate_iterator& operator=(enumerate_iterator const&) = default;
+  enumerate_iterator& operator=(enumerate_iterator&&) noexcept = default;
+
+  ~enumerate_iterator() noexcept = default;
+
+  friend void swap<>(enumerate_iterator<T, C>& a, enumerate_iterator<T, C>& b);
+
+  using difference_type = typename std::iterator_traits<T>::difference_type;
+  using value_type = std::pair<C, typename std::iterator_traits<T>::value_type>;
+  using pointer = std::pair<C, typename std::iterator_traits<T>::pointer>;
+  using reference = std::pair<C, typename std::iterator_traits<T>::reference>;
+  using iterator_category = typename std::iterator_traits<T>::iterator_category;
+
+  reference operator*() const;
+
+  enumerate_iterator& operator++();
+
+  bool operator==(enumerate_iterator const& other) const;
+  bool operator!=(enumerate_iterator const& other) const;
+
+  T _iter;
+  C _c;
+};
+
+template <typename T, typename C>
+auto enumerate_iterator<T, C>::operator*() const -> reference {
+  return std::make_pair(_c, std::reference_wrapper(*_iter));
+}
+
+template <typename T, typename C>
+struct enumerate_wrapper {
+  using I = typename std::decay_t<T>::iterator;
+  T _t;
+  C _c;
+
+  auto begin() { return enumerate_iterator(_t.begin(), _c); }
+  auto end() { return enumerate_iterator(_t.end(), _c); }
+};
+
+template <typename T, typename C>
+void swap(enumerate_iterator<T, C>& a, enumerate_iterator<T, C>& b) {
+  using std::swap;
+  swap(a._iter, b._iter);
+  swap(a._c, b._c);
+}
+
+template <typename T, typename C>
+auto enumerate_iterator<T, C>::operator++() -> enumerate_iterator& {
+  ++_c;
+  ++_iter;
+  return *this;
+}
+
+template <typename T, typename C>
+bool enumerate_iterator<T, C>::operator==(enumerate_iterator const& other) const {
+  return _iter == other._iter;
+}
+
+template <typename T, typename C>
+bool enumerate_iterator<T, C>::operator!=(enumerate_iterator const& other) const {
+  return _iter != other._iter;
```

Comment:
! (*this == other)
```suggestion
  return !( *this == other);
```

Reply:
Why? This would replace the logic of the underlying iterator.

Reply:
Then the implementation of the underlying container is broken. Lets assume `!=` is faster to detect for the underlying container than `==` then `a == b` should probably be implemented as `!(a != b)` there. I prefer to reuse already implemented functions instead of going down to lower implementation levels several times. But it is just a matter of taste and not really important.